res/compile.py

```diff
@@ -0,0 +1,193 @@
+inclex = re.compile('#include\s+"(.*)"')
+commex = re.compile("//.*?\n")
+
+def addCRC(filename):
+    with open(filename) as file:
+        file = open(filename,'r+b')
+        buf = file.read()
+        crc = (binascii.crc32(buf) & 0xFFFFFFFF)
+        file.seek(0)
+        file.write(struct.pack("<I", crc))
+        file.write(buf)
+
+def findOptions(shader):
+    list = []
+    with open(shader) as f:
+        buffer = f.read()
+        buffer = re.sub(commex, "", buffer) # remove commented lines
+        for option in options:
+            if option in buffer:
+                list.append(option)
+        includes = inclex.findall(buffer)  # find all include files
+        for include in includes:
+            list += findOptions(os.path.join(os.path.dirname(shader), include))
+    return list
+
+def findEntry(shader):
+    with open(shader) as f:
+        buffer = f.read()
+        buffer = re.sub(commex, "", buffer) # remove commented lines
+        match = entrex.search(buffer)  # find the entry point
+        if match:
+            return match.group(0)
+        if 'main' in buffer:
+            return 'main'
+        includes = inclex.findall(buffer)  # find all include files
+        for include in includes:
+            if include != 'common.h':
+                return findEntry(os.path.join(os.path.dirname(shader), include))
+    return False
+
+def generateIds(list, id, i):
+    if i >= len(options):
+        return [ id ]
+    if options[i] in list:
+        if options[i] in forceOptions:
+            return generateIds(list, id + '1', i + 1)
+        if options[i] in disableOptions:
+            return generateIds(list, id + '0', i + 1)
+        if options[i] in skinOptions:
+            ids = []
+            for (index, option) in enumerate(skinOptions):
+                skin = '0' * len(skinOptions)
+                skin = skin[:index] + '1' + skin[index+1:]
+                ids += generateIds(list, id + skin, i + len(skinOptions))
+            return ids
+        if options[i] in qualityOptions:
+            ids = []
+            for index in range(0, qualityOptions[options[i]]):
+                ids += generateIds(list, id + str(index), i + 1)
+            return ids
+        return generateIds(list, id + '0', i + 1) + generateIds(list, id + '1', i + 1)
+    else:
+        return generateIds(list, id + '_', i + 1)
+
+def compileShader(name, stage, flags):
+    path = os.path.join(shaders, '%s.%s' % (name, stage))
+    entry = findEntry(path)
+    if not entry:
+        print('No entry point found, skipping shader...')
+        return
+
+    profile = '%s_%s' % (stage, featureLevel)
+    if len(entry) > len('main'):
+        profile = entry[5:]
+    
+    object = os.path.join(objects, '%s.%s' % (name, stage))
+    if os.path.isfile(object):
+        print('Already compiled, skipping object %s' % flags)
+        return
+    
+    args = [fxc, '/nologo', '/LD', '/Zpr', '/E%s' % entry, '/T%s' % profile]
```

Comment:
/O1 is the default optimization level and any level above that is reserved for future use. The reason for not specifying the optimization level is because some shaders use old 1.x shader models which don't support any optimization level.